### PR TITLE
Shared CLJ/CLJS lib 

### PR DESCRIPTION
**Pull request by camsaul**
Creates a shared Clojure/ClojureScript library. As a proof of concept, this PR moves the type hierarchy into a file shared between both frontend and backend. Previously, the frontend had to fetch the type hierarchy via an API request and then manually rebuild a hierarchy object and an `isa?` function. This PR eliminates a lot of that duplicate code from `types.js`.

Build scripts, etc. work without any noticeable difference to anyone using them, although you have to have Java installed now to compile the shared CLJS code. I think this is an acceptable requirement because you can't compile the rest of Metabase or run it without having Java anyway

#### DEMO

Here's an older demo of the shared lib. I'm using a shared `.cljc` file between both the backend and frontend. This demo uses it in two places: one in the backend to generate the X-Ray titles, and another in the frontend to generate the `Hello, Cam` greeting at the top.

Watch as I edit the source and swap out the emoji it adds to the strings. Hot-reloading in both frontend and backend.

liveReloading

The code demoed isn't included in this PR -- it was just for demo purposes. The code in this PR actually does something useful now

#### MORE INFO

- Uses https://github.com/thheller/shadow-cljs
- Takes inspiration from #2573 and #9157
- Hot-loading fully working
- Optimized/minified build of shared CLJS code is only 144k. This is before any compression takes place

#### TODO

- [ ] fix the `Namespace "goog.math.Long" already declared.` error in tests
- [x] Prettier should skip the generated files in `frontend/src/cljs`

### CODE REVIEW

**Comment by dpsutton** [comment]
> I haven't integrated this into the main build script yet. When I do, we can use the Closure compiler to do aggressive dead-code elimination and keep the added code size tiny

You are running `shadow-cljs watch app` I'm guessing? You should be able to get a good idea of size by just running `shadow-cljs release app` and checking the size of the output. I'm a bit confused by the output directory. Is the output into the source tree? `:output-dir "../frontend/src/cljs/"`?

**Comment by dpsutton** [comment]
Also, i'm surprised to see the shadow config located under /shared and not top level. this might get weird in the sense that we now have a second package.json file (in addition to a gitignore file. I was expected to see it top level so that it has access to react and all the other goodies we may want (unless npm is aware of a top level package.json file?)

**Comment by camsaul** [comment]
I'm not really sure what the best way to organize this is. Top-level using the shared package.json sounds like it would make sense

**Comment by camsaul** [comment]
@dpsutton have you worked on mixed projects like this before? I haven't really spent that much time playing with ClojureScript, so I could use some help trying to figure out how to set it up

**Comment by dpsutton** [comment]
never on a mixed project. always figwheel main driving it. I'm hoping just putting this top level will work. shadow can choke on some more modern and tool-heavy js stuff but since we're emitting npm-module it might be a bit simpler. we do have to watch bundle sizes though if that's a concern. we can optimize a bunch away but always have to pay for the cljs runtime and all the immutable data structures. I can look at it tomorrow. were you able to get a repl up in CIDER?

**Comment by camsaul** [comment]
I was able to get either a a Clojure or ClojureScript REPL for that code up in CIDER, but didn't try to get a REPL that was connected to the app in the browser in some way

**Comment by camsaul** [comment]
Update: I've made some changes after feedback from @dpsutton. 

- CLJS code now shares a `package.json` file with the main JS FE code. 
- Tweaks various commands like `yarn build` and `yarn run build-hot` to build both CLJS and JS (ported from #14901 
- Production build when running `./bin/build` or the like -- I've measured this and from what I can tell it only adds around 100k to the bundle size. It's pretty good at eliminating the unused code :shrug:
- Moved `types.clj` to the shared library; reworked `types.js` to use the shared code. 

Everyting seems like it works for me locally but there are still a few things that need to be ironed out. FE unit tests fail with

```
Summary of all failing tests
 FAIL  frontend/test/********/containers/SavedQuestionLoader.unit.spec.js
  ● Test suite failed to run

    Namespace "goog.math.Long" already declared.
      
      at Object.<anonymous>.goog.module (frontend/src/cljs/cljs_env.js:124:13)
      at frontend/src/cljs/goog.math.long.js:14:8
      at Object.<anonymous>.goog.loadModule (frontend/src/cljs/cljs_env.js:322:27)
      at Object.<anonymous> (frontend/src/cljs/goog.math.long.js:12:6)
      at Object.<anonymous> (frontend/src/cljs/cljs.core.js:4:1)
      at Object.<anonymous> (frontend/src/cljs/********.types.js:4:1)
```

GitHub actions still need to be tweaked to build the CLJS code before running FE actions. Cypress tests are failing -- haven't looked in to reasons why yet.

**Comment by camsaul** [comment]
@paulrosenzweig @ariya do you guys have any ideas as to why the 

```
Summary of all failing tests
 FAIL  frontend/test/********/containers/SavedQuestionLoader.unit.spec.js
  ● Test suite failed to run

    Namespace "goog.math.Long" already declared.
      
      at Object.<anonymous>.goog.module (frontend/src/cljs/cljs_env.js:124:13)
      at frontend/src/cljs/goog.math.long.js:14:8
      at Object.<anonymous>.goog.loadModule (frontend/src/cljs/cljs_env.js:322:27)
      at Object.<anonymous> (frontend/src/cljs/goog.math.long.js:12:6)
      at Object.<anonymous> (frontend/src/cljs/cljs.core.js:4:1)
      at Object.<anonymous> (frontend/src/cljs/********.types.js:4:1)
```

error would be happening?

**Comment by dpsutton** [comment]
that's gonna be purely from the goog stuff that cljs and closure library provide so i doubt they will know.

**Comment by dpsutton** [comment]
I just pulled this down and i'm getting an error running `shadow-cljs compile app`:

```
Invalid configuration
-- Spec failed --------------------

  {:target :npm-module, :output-dir "frontend/src/cljs/", :build-id :app}

should contain key: :entries

-- Relevant specs -------

:shadow.build.targets.npm-module/target:
  (clojure.spec.alpha/keys
   :req-un
   [:shadow.build.targets.shared/output-dir
    :shadow.build.targets.npm-module/entries]
   :opt-un
   [:shadow.build.targets.npm-module/runtime
    :shadow.build.targets.shared/devtools])
:shadow.build.config/build+target:
  (clojure.spec.alpha/and
   :shadow.build.config/build
   (clojure.spec.alpha/multi-spec
    shadow.build.config/target-spec
    :target))
```

from [the docs](https://shadow-cljs.github.io/docs/UsersGuide.html#target-npm-module) we need some entries

seems like beyond enumerating the namespaces to compile, the output isn't all rosy either:

> If you use the default :output-dir of "node_modules/shadow-cljs" you can access the declared namespaces by using require("shadow-cljs/demo.foo") in JS. When using something not in node_modules you must include them using a relative path. With :output-dir "out" that would be require("./out/demo.foo") from your project root.

**Comment by dpsutton** [comment]
Also, things are gonna fail under advanced compilation because closure renames absolutely everything. To combat this we have to annotate the exported vars with `^:export` metadata


```
diff --git a/frontend/src/metabase/lib/greeting.js b/frontend/src/metabase/lib/greeting.js
index abba71aad..799a188cc 100644
--- a/frontend/src/metabase/lib/greeting.js
+++ b/frontend/src/metabase/lib/greeting.js
@@ -1,4 +1,5 @@
 import { t } from "ttag";
+import { greeting } from "cljs/metabase.types";
 
 const greetingPrefixes = [
   t`Hey there`,
@@ -20,7 +21,8 @@ const Greeting = {
     const randomIndex = Math.floor(
       Math.random() * (greetingPrefixes.length - 1),
     );
-    return greetingPrefixes[randomIndex];
+      //return greetingPrefixes[randomIndex];
+      return greeting();
   },
 
   sayHello: function(personalization) {
diff --git a/shadow-cljs.edn b/shadow-cljs.edn
index 9ba6e53ca..3b2db0a92 100644
--- a/shadow-cljs.edn
+++ b/shadow-cljs.edn
@@ -7,4 +7,5 @@
 
  :builds
  {:app {:target     :npm-module
+        :entries [metabase.types]
         :output-dir "frontend/src/cljs/"}}}
diff --git a/shared/src/metabase/types.cljc b/shared/src/metabase/types.cljc
index ef4035563..6e7c57e1b 100644
--- a/shared/src/metabase/types.cljc
+++ b/shared/src/metabase/types.cljc
@@ -239,15 +239,18 @@
   (some #(isa? % :type/Temporal) [base-type semantic-type]))
 
 #?(:cljs
-   (defn isa
+   (defn ^:export isa
      "Is `x` the same as, or a descendant type of, `y`?"
      [x y]
      (isa? (keyword x) (keyword y))))
 
 #?(:cljs
-   (def TYPE
+   (def ^:export TYPE
      "A map of Type name (as string, without `:type/` namespace) -> qualified type name as string
 
          {\"Temporal\" \"type/Temporal\", ...}"
      (clj->js (into {} (for [tyype (descendants :type/*)]
                          [(name tyype) (shared.u/qualified-name tyype)])))))
+
+(defn ^:export greeting []
+  (rand-nth ["yee-haw" "podnuh" "what's it do"]))

```
this patch kinda does what needs to be done. i added a little greeting and threw it in there just to make sure everything worked.

An easy thing to do while you are testing is to run `npx shadow-cljs release app` after the watch compiles after changes. Shadow is awesome and this should be pretty quick since it sees the server already running so there's no jvm startup. This will mangle, munge, and concatenate everything through closure so its exactly what a prod build will produce.

So downsides:
1. need to enumerate exported namespaces in `:entries` vector in `shadow-cljs.edn`
2. need to ensure that we mark exported vars with `^:export` otherwise it will fail on production builds and we won't see it until its too late

I'm hoping the naming stuff is the source of woe for all of the weirdness. I still don't understand why you were able to build anything with what my version of shadow complained was an invalid config file.

**Comment by dpsutton** [comment]
image

adding reagent and

```clojure
(ns metabase.component
  (:require [reagent.core :as r]))

(defn ^:export silly
  []
  (r/as-element [:div
                 [:h3 "this is some react stuff"]
                 [:h4 "and some more"]
                 [:p "and this is text for stuff"]]))
```

Not that we want to do components in cljs but we could. And it does seem like we will need to enumerate the namespaces. That's a curious thing.

**Comment by dpsutton** [comment]
Update: talked with thheller over on clojure slack, and it seems like we can do `:entries []` and we'll be fine, it will compile all namespaces. so that solves one problem. so the major footgun remaining is the munging from names requiring `^:export`. I asked if there was a way to fix that, hoping we could get munging in dev so lack of export was obvious. Unfortunately that's provided by google closure so only available in release builds.

**Comment by camsaul** [comment]
@dpsutton I thought I read somewhere that `^:export` wasn't required anymore for newer versions CLJS. It is munging the names tho I guess. I'm reading the shadow-cljs dox on generating NPM modules https://shadow-cljs.github.io/docs/UsersGuide.html#target-npm-module to wrap my head around it a bit more.

You can use the `yarn` aliases to build the CLJS code. `yarn build`, `yarn build-hot`, and `yarn build-watch` all now build both the JS and CLJS code. You can do `yarn build[-whatever]:cljs` or `:js`  to do just one or the other, but the un-suffixed versions do both. Not sure why `npx shadow-cljs` wasn't working for you, but I didn't try that.

**Comment by camsaul** [comment]
The version of shadow-cljs I pinned in `package.json` doesn't seem to care whether or not `:entries` is listed for `:npm-module`... it seems like it does the same thing whether or not it's there. The dox however do say it's required: https://shadow-cljs.github.io/docs/UsersGuide.html#target-npm-module

So maybe that requirement was added (or removed?) recently

File: src/metabase/public_settings.clj [hunk 1]
```diff
@@ -323,18 +322,6 @@
...
-  "Field types"
-  :visibility :public
-  :setter     :none
-  :getter     (fn [] (types/types->parents :type/*)))
-
-(defsetting entities
-  "Entity types"
-  :visibility :public
-  :setter     :none
-  :getter     (fn [] (types/types->parents :entity/*)))
```
Comment (paulrosenzweig) [inline src/metabase/public_settings.clj:336]: I guess we were already not using this?

**Review by paulrosenzweig — APPROVED** [review]
Looks good! It's a fair amount of additional complication to the FE build, but that'll be worth it when more code is moved here.

I find the use of "shared" in both `shared/src` and `frontend/src/metabase-shared` confusing. This is more truly "shared", so maybe `frontend/src/metabase-shared` should be renamed "pulse rendering" or something.

File: src/metabase/public_settings.clj (hunk 1 above)
Comment (camsaul) [reply src/metabase/public_settings.clj:336]: Yeah, I could not find any place that was using this. Removing it didn't affect anything at least

**Comment by ariya** [comment]
JFYI, since we ought to pay attention to the size of the front-end assets (to prevent some terrible slow-down like #13472): apparently this shared CLJ/CLJS lib added **only** about 27 KB to `app-main.bundle.js` (release build, included in the Uberjar).

Subjectively, this does **not** incur any noticeable (cold start, no cache) performance impact. Thus, for the benefits that we get from sharing the code, IMO this is worth the extra 27 KB!

**Pull request merged**